### DISCUSSION TITLE
Using PP as a base for embroidery tool?

### DISCUSSION THREAD

**Discussion by hluedeke** [Pattern-Projector/pattern-projector#300] [Ideas]
Ok, don't hate me. I just had this random idea today. I bet a lot of people who would use Pattern Projector for cutting out patterns could also use a similar product and/or feature for embroidery.

In today's world, it's so hard to determine embroidery placement on stuff. You can do a "best guess" based on measurements, you can print out paper copies and arrange stuff, etc, but wouldn't it be really cool to have a similar setup to PP (although maybe branded differently) where you calibrate and then can upload one or more embroidery files for visualization? You could then project your "embroidery playground" onto your piece of clothing or whatever and mess around with design placement, rotation, etc.

There would be crosshair displayed visually from the software (and probably a visual "hoop") so that you could mark your center line(s) for placement. You could even potentially plan multi-hoop larger designs with some well thought-out tools.

This is totally half-baked, and I'm not really asking anyone for anything. I'm certainly not asking for PP to become an embroidery software. I just wanted to throw it out there while the inspiration was fresh! It just seems like this projection calibration technology could be really helpful in this area as well.

**Comment by courtneypattison** [comment]
This is a really cool idea! I've done two embroidery projects so I have very little knowledge of this area. I do want to try and do more though. I added some daisies to a dress I made for my sisters wedding a couple of weeks ago and it looks so cute!

What's the typical file format for embroidery?

I've used PP for quite a few non sewing projects like cardboard cutting for masks and making a fairy door (cuts lines for the wood and pattern for painting). Maybe it would make sense to have different modes for different use cases. Like an embroidery, painting, cardboard cutting, and sewing mode where each one shows different tools that make sense for that use. Mostly I've been trying to add things that I can see being useful across all these cases so far.

**Reply by hluedeke** [reply]
> Mostly I've been trying to add things that I can see being useful across all these cases so far.

Yeah, and that totally makes sense. It's really the simplest and most straightforward option.

I don't know what the most used format is, but I'd *guess it would be a PES file (made by Brother). There are a lot of tools (including free ones I think) that convert between file formats, so I would probably start with just the one. I've already actually got a parser working locally on my machine (thank you open source and ChatGPT!), but it will take more effort to write to the screen as I don't think there are any NodeJS display libraries out there.

I think my initial focus would be just display in a basic, non-transformed canvas. Once I felt confident with that, I would work toward integrating it to a calibration canvas. I think the biggest differences would be potential multiple file uploads and multiple manipulative objects. The other thing would be having a "center crosshair" or a "hoop" artifact on the canvas to help with positioning.

I also think I should research the relationship between "center of the hoop" and how the file loads into a machine - to be sure that what you see on the projector is what you get from the machine.

It's an interesting idea at least. If we did do different use case toolsets, maybe we keep them in the codebase monolith but build & brand each one separately - like embroidery would be "Embroidery Projector", maybe with the same logo in a different color or something like that.

I don't really know. Just dreaming! I'm glad you at least see the usefulness in it. I'm not even sure if it's feasible at this point, but I might do some tinkering to see if I can come up with something.

**Comment by hluedeke** [comment]
<img width="912" alt="its-a-start" src="https://github.com/Pattern-Projector/pattern-projector/assets/14338579/9242cb12-6900-4ac8-8131-00f656c9d8ad">

Whelp, it's a start at least 😆 it's supposed to be a pug. Actually, I think all I'm missing is the transform on a color change. So I'm rather proud of this progress, for what it's worth! It's awesome to "crack the code" and be able to draw the file to a canvas. So cool!

**Reply by courtneypattison** [reply]
This looks awesome! You're getting me excited about embroidery!

**Comment by hluedeke** [comment]
Getting there:

<img width="601" alt="image" src="https://github.com/Pattern-Projector/pattern-projector/assets/14338579/4901c3e7-945f-4aa9-8db7-7aee67c3a7de">


<img width="491" alt="image" src="https://github.com/Pattern-Projector/pattern-projector/assets/14338579/a582e3c5-f98a-49cf-8de6-34261ad71451">

It's a lot of trial and error, but also super cool. Hopefully I can pinpoint all the minor tweaks and having something plottable soon!

**Reply by courtneypattison** [reply]
This looks so cool. Have you tried projecting it yet? I'd love to see what it looks like, even without the transform!

**Reply by hluedeke** [reply]
I have not - I am finding my life is full of distractions right now 😂 so I'm getting to it when I can! It would be cool to try though!

**Reply by courtneypattison** [reply]
I'm in the same boat!!! This release will get out eventually lol.

**Comment by hluedeke** [comment]
<img width="1196" alt="image" src="https://github.com/Pattern-Projector/pattern-projector/assets/14338579/c482192b-89f0-4eb2-b419-c2da5e8eb148">


This is still very rough, but I decided to start integrating it into the main pattern projector tool (I had to do some small modifications to the main tool to support either PDF patterns OR embroidery files). I got it to the point where it's drawing to the screen on the calibration transform. I have not checked to see if it is accurate yet, but it passes the "common sense" size test. There is also a stitch in there that I am not reading correctly, but I think that will be a small change when I decide to go tackle it. It will just be tedious.

I've added support for transforming multiple objects on the screen instead of just one so that you can upload more than one file at a time. I've also drawn in crosshairs/a border to simulate a hoop with centered lines. The idea is that you can layout your designs then you can mark the center hoop lines on your fabric from the projector. Heck, you might even be able to mark the outer hoop lines as a sanity check.

Anyway, the toolbar is going to be completely different for embroidery vs. pdf patterns. Not only should you be able to move and rotate hoops to any location or angle, but eventually it would be great to allow transforms inside the hoop (this is useful for more than one design in the same hooping). I think those transformations would be handled by user input fields instead of dragging, and the hoop itself would be controlled more by dragging/handles, etc. Those transformations inside the hoop would have to be very visible (as lists of numbers) as well, as I'm not sure if it's possible (without merging design files) to apply transformations outside of manually doing it on your machine.

Obviously this isn't perfect for everything, but I'm thinking it will be great for:

1) sanity checking simple flat designs, like on a t-shirt for a size/placement reference
2) for large or repeating designs across a wide area (I want to use it for a table runner for instance).

A few concerns I have:

1) Accuracy. We always say "it's not that important to be accurate when cutting", but it's kinda more important to be accurate with embroidering. Maybe this will be fine, but I'm a little worried that designs won't be perfectly centered because of fabric distortion, even with "perfect" calibration on the screen.
2) Non-flat items. I would love to embroider the hem of a skirt with something like this, but if it's not perfectly flat, I don't think this tool would have much merit. You would definitely want to embroider before you attached/ruffled a skirt vs. after

**Reply by hluedeke** [reply]
Other really awesome potential features:
- Duplicate. Copy/paste the same design over and over for quick repeats
- Layout functions. Select multiple items and equally space them apart or align them vertically.

**Reply by courtneypattison** [reply]
Ooo I'm getting excited to try this out!

I'm kind of thinking separating this tool out into its own app would make the most sense (I'm pretty sure you mentioned this somewhere). Even just embroidery.patternprojector.com because it does have quite a few unique features, even just the file format. What do you think? There would be some code duplication but we could work toward making a shared library to use between the two. I don't want your design choices to be stifled by the limitations of PP.

I'm hopeful that your accuracy concerns won't be that big of a deal! I think a few pattern weights and careful calibration should do the trick but we'll have to test to see.

**Reply by hluedeke** [reply]
Yeah, agree. I would love to obviously "share" the underlying calibration grid code, as that will remain unchanged. But overall it is going to start diverging quickly. It's already on my fork in its own branch, so that branch will probably end up as its own named project.

I'm hoping so on accuracy too. I should probably pause and test it out but where's the fun in that 🤪

I have been traveling quite a bit lately and will be traveling again over the next few weeks, so this will be slow going as usual. Let me know if there is anything on the main project you want me to take a look at as well. I feel like I've just kinda gone down a rabbit hole with this, so I'm happy to pause and go back to PP if needed!

**Reply by courtneypattison** [reply]
No need to go back to PP! Have fun with this project! I've been prioritizing other things lately as well :)

**Reply by hluedeke** [reply]
I thought I'd come back in here just to state the obvious: I have not gone anywhere with this idea 🤣 life happens man. This might be a fun retirement project or something someday haha. I love the idea of it! Someday, someday

**Reply by courtneypattison** [reply]
Lol, I feel you! My computer and craft room are littered with half finished projects!
